### ISSUE TITLE
Start reading from the current viewpoint in the Browser

### ISSUE THREAD

**Issue by Stvad** [jamestomasino/stutter#40]
Hey, it'd be awesome if I'd be able to (optionally?) initiate stutter reading starting at the exact text I'm looking at and not from the beginning of the page. 

Also is there any way at all to skip to a point in text that I'm interested in? (i.e. if I'm resuming reading a large article that I started on before). Besides selecting huge swath of text, I guess 😛?

**Comment by jamestomasino** [comment]
**TL;DR** - It's extremely difficult to do but could be possible if some other hurdles are tackled first. If we do it, we'll need to add a new context-menu option.

-----

That's a good question. I'm trying to think of this in two ways:

1) is it possible
2) should we do it

For question 1, I would need to identify what the top-most visible element on screen is to figure out where to start pulling text. Of course, you could be in the midst of a long block, so I'd need to look at the text itself to see if it's on screen. There's no good way to do that, but [there's some bad ones](https://stackoverflow.com/questions/19356819/how-to-find-the-last-visible-word-position-in-a-text-container). Most involve cloning the block and systematically erasing it until the container boundary passes some test, like a height change. Not great.

The other difficulty in question 1 is that we're using the Mozilla Readability library to get clean text for display. This only runs when using the icon to read the "full page" currently, because I needed all their logic to parse what's real and what's sidebar/menu/nav/ads/etc. I'd still need that in this case, but I'd have to find a way to link it to the screen. That's actually already part of issue #15. If we were able to link the text being read to the screen, this would be significantly easier. I don't see how to do it with the current architecture, though.

Then there's question 2. Should we do it? I think most would agree that's not the expected behavior of the icon, so we'd have to add a different UI trigger. Perhaps another option in the right click context menu could be "read from here". That might help a bit with question 1 if the context menu somehow identified the word to start on. Linking it all together is still a problem.

**Comment by jamestomasino** [comment]
#15

**Comment by jamestomasino** [comment]
Oh, one other thing. It's not the perfect solution, but we did add keyboard shortcuts recently that might help. If you're restarting a page and need to fast forward to where you were, you can use Alt+RightArrow and Alt+LeftArrow to scrub in the playback timeline. Alt+Up/Down will increase or decrease speed of playback, and the "Enter" key will pause.

**Comment by Stvad** [comment]
Thanks for a swift and detailed reply! :) 

Doing things based on element visibility could be a good first step towards this. As you mentioned long blocks can be a problem, but in general I feel like it's better to start from the start of the block then from the start of the page for the `resume` reading scenario.

As potential source of inspiration for determining where you are - you can look at https://github.com/brookhong/Surfingkeys plugin. It can create visual hints for navigation purposes for each word displayed on the screen. Not sure about the mechanism used (maybe they use the one you referenced) but something to look at. The performance of hint creation in it is pretty good.

I don't think that doing this requires solving back-linking problem. Only initial discovery of starting position, which seems like a simpler problem to solve.

One simple mechanics that comes to mind (though I suspect a not very intuitive one from usability standpoint) is to ask user to select a few consecutive words and then click "start reading from here"  in that case you can treat those words as an identifier of position in text (i.e. search the text you get after extraction till you find match, start reading from there)

---

One might argue expected behavior either way ;) , a separate trigger (preferably via keyboard shortcut in addition to context menu)  should work just fine and may potentially help solve other problems as discussed before.

**Comment by Stvad** [comment]
Fast forward shortcuts are helpful, thank you! Is there one to trigger the stutter in a first place? :)

**Comment by jamestomasino** [comment]
> Fast forward shortcuts are helpful, thank you! Is there one to trigger the stutter in a first place? :)

Not at the moment. Webextensions give a mechanism for this in the manifest file but when I tried to implement it nothing happened. It's probably worth another try.

**Comment by jamestomasino** [comment]
Alt-Shift-R will now start Stuttering! Hooray for global hotkey.

**Comment by Stvad** [comment]
😿

**Comment by jamestomasino** [comment]
You can select text first before stuttering to read only that portion. It's probably the best option we're going to get until text fragment api becomes standard
